### PR TITLE
Skip adding/accumulating ClipRect values which corresponding rect offset is auto.

### PR DESCRIPTION
**Pull request by mantaroh**
This is a PR for https://bugzilla.mozilla.org/show_bug.cgi?id=1390352

This patch will skip adding/accumulating the values which corresponding rect offset is auto, and  make Servo_AnimationValues_ComputeDistance return negative value instead of 0.0 when the function fails to distinguish its failure.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

There are already these tests in dom/smil/tests of gecko, this PR will enable these tests.
For detail, see https://bugzilla.mozilla.org/show_bug.cgi?id=1390352.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17288)

### CODE REVIEW

**Comment by mantaroh** [comment]
This patche have been reviewed by @hiikezoe , @Manishearth , @birtles 
(https://reviewboard.mozilla.org/r/169612/ and https://reviewboard.mozilla.org/r/171008)

**Comment by mantaroh** [comment]
```
  ▶ TIMEOUT [expected OK] /_mozilla/mozilla/paint_timing.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 17.2.0-devel

  ▶ Unexpected subtest result in /_mozilla/mozilla/paint_timing.html:
  └ NOTRUN [expected PASS] Performance entries observer
```

@jdm 
I can't find this intermittent from issues.
Do you know this type error?

File: ports/geckolib/glue.rs [hunk 1]
```diff
@@ -384,7 +384,7 @@ pub extern "C" fn Servo_AnimationValues_ComputeDistance(from: RawServoAnimationV
                                                         -> f64 {
     let from_value = AnimationValue::as_arc(&from);
     let to_value = AnimationValue::as_arc(&to);
-    from_value.compute_squared_distance(to_value).map(|d| d.sqrt()).unwrap_or(0.0)
+    from_value.compute_squared_distance(to_value).map(|d| d.sqrt()).unwrap_or(-1.0)
```
Comment (nox) [inline ports/geckolib/glue.rs]: You explained the change in the commit, but a comment right here would be useful to convey that a negative distance means an error occurred.

File: components/style/properties/helpers/animated_properties.mako.rs [hunk 2]
```diff
@@ -1106,6 +1106,22 @@ impl Into<FontStretch> for f64 {
 impl<H, V> RepeatableListAnimatable for generic_position::Position<H, V>
     where H: RepeatableListAnimatable, V: RepeatableListAnimatable {}
 
+/// https://drafts.csswg.org/css-transitions/#animtype-rect
+impl Animate for ClipRect {
+    #[inline]
+    fn animate(&self, other: &Self, procedure: Procedure) -> Result<Self, ()> {
+        if !self.can_animate_with(other, procedure) {
+            return Err(());
+        }
```
Comment (nox) [inline components/style/properties/helpers/animated_properties.mako.rs]: There is a better way to do that:

```rust
let animate_component = |this, other| {
    match (this.animate(other, procedure)?, procedure) {
        (None, Procedure::Interpolate { .. }) => Ok(None),
        (None, _) => Err(()),
        (result, _) => Ok(result),
    }
}
```

This way you don't need code to check whether the two values should be allowed to interpolate at all beforehand.

You could even wrap the `Option<T>` of the `ClipRect` fields in its own new type and implement `Animate` by hand for it, keeping the derived code for `ClipRect`.

**Review by nox — CHANGES_REQUESTED** [review]

**Comment by jdm** [comment]
@mantaroh That's #18219; it's a new test that was just added yesterday.

**Comment by mantaroh** [comment]
Thanks @jdm !
I'll retry after addressing it.

File: components/style/properties/helpers/animated_properties.mako.rs (hunk 2 above)
Comment (mantaroh) [reply components/style/properties/helpers/animated_properties.mako.rs]: Thanks.
Ah! OK. I'll use Closure!

**Comment by upsuper** [comment]
I think you can land this PR with previous review from Bugzilla as far as @nox doesn't complain about the updated patch.

**Comment by nox** [comment]
Would have been nice to have a better spec link on the `Animate` impl, the one you put doesn't show that animating `auto` with `auto` should only work when interpolating.

**Review by nox — APPROVED** [review]

**Comment by jdm** [comment]
That failure was #13480, so you can retry.

**Comment by jdm** [comment]
```
   Compiling mime_guess v1.8.1
error: Could not compile `mime_guess`.
```
Not sure what to make of this error, but it's not related to these changes; probably need to investigate the build machine.

**Comment by jdm** [comment]
In any case, go ahead and retry.

**Comment by mantaroh** [comment]
@nox , @jdm 
Many thanks.

**Comment by mantaroh** [comment]
I think test fail of mac-rel-wpt1 is related to #18228
@bors-servo retry

**Comment by mantaroh** [comment]
Thanks @upsuper !

**Pull request merged**